clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy

```diff
@@ -16,203 +16,391 @@
-    virtualIP.poolId = resultPool.id
-    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Created lbPool ${newLoadBalancerPool.name} in ${region}"
+      if (listenersToUpdate) {
+        updateListenersAndPools(region, resultLoadBalancer.id, description.algorithm, listenersToUpdate.values(), description.healthMonitor)
+      }
 
-    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Creating vip for lbPool ${resultPool.name} in ${region}..."
-    Vip vip = openstackClientProvider.createVip(region, virtualIP)
-    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Created vip for lbPool ${resultPool.name} in ${region} with name ${vip}."
+      updateFloatingIp(region, description.networkId, resultLoadBalancer.vipPortId)
+      updateSecurityGroups(region, resultLoadBalancer.vipPortId, description.securityGroups)
 
-    if (poolHealthMonitor) {
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Creating health checks for lbPool ${resultPool.name} in ${region}..."
-      openstackClientProvider.createHealthCheckForPool(region, resultPool.id, poolHealthMonitor)
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Created health checks for lbPool ${resultPool.name} in ${region}."
+      // Add members to newly created pools through an existing stack only
+      if (description.id && (!listenersToAdd.isEmpty() || !listenersToDelete.isEmpty())) {
+        updateStacks(region, resultLoadBalancer.id)
+      }
+    } catch (OpenstackProviderException ope) {
+      throw new OpenstackOperationException(AtomicOperations.UPSERT_LOAD_BALANCER, ope)
     }
 
-    if (description.networkId) {
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Obtaining network name from network id $description.networkId..."
-      Network network = openstackClientProvider.getNetwork(region, description.networkId)
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Successfully obtained network name from network id $description.networkId."
+    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Done upserting load balancer ${resultLoadBalancer?.name} in ${region}"
+    return [(region): [id: resultLoadBalancer?.id]]
+  }
 
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Obtaining floating IP from network $network.name..."
-      FloatingIP ip = openstackClientProvider.getOrCreateFloatingIp(region, network.name)
-      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Successfully obtained floating IP from network $network.name."
+  /**
+   * Creates a load balancer in given subnet.
+   * @param region
+   * @param name
+   * @param subnetId
+     * @return
+     */
+  protected LoadBalancerV2 createLoadBalancer(String region, String name, String subnetId) {
+    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Creating load balancer $name in ${region} ..."
+    String description = generateCreatedTime(System.currentTimeMillis())
+    LoadBalancerV2 result = BlockingClientAdapter.from { LoadBalancerV2 loadBalancer ->
```

Comment:
This is a bit confusing - I was expecting the from to be the operation to execute, and the execute (check) to be the check to run on.